specification/hybridcompute/resource-manager/Microsoft.HybridCompute/preview/2022-08-11-preview/HybridCompute.json

```diff
@@ -0,0 +1,1553 @@
+        "responses": {
+          "200": {
+            "description": "OK",
+            "schema": {
+              "$ref": "#/definitions/MachineExtensionsListResult"
+            }
+          },
+          "default": {
+            "description": "Error response describing why the operation failed.",
+            "schema": {
+              "$ref": "../../../../../common-types/resource-management/v2/types.json#/definitions/ErrorResponse"
+            }
+          }
+        },
+        "x-ms-pageable": {
+          "nextLinkName": "nextLink"
+        },
+        "x-ms-examples": {
+          "GET all Machine Extensions": {
+            "$ref": "./examples/LISTExtension.json"
+          }
+        }
+      }
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.HybridCompute/machines/{machineName}/upgradeExtensions": {
+      "post": {
+        "tags": [
+          "MachineExtensions Upgrade"
+        ],
+        "operationId": "UpgradeExtensions",
+        "description": "The operation to Upgrade Machine Extensions.",
+        "parameters": [
+          {
+            "$ref": "../../../../../common-types/resource-management/v1/types.json#/parameters/ApiVersionParameter"
+          },
+          {
+            "$ref": "../../../../../common-types/resource-management/v1/types.json#/parameters/SubscriptionIdParameter"
+          },
+          {
+            "$ref": "../../../../../common-types/resource-management/v1/types.json#/parameters/ResourceGroupNameParameter"
+          },
+          {
+            "name": "machineName",
+            "in": "path",
+            "required": true,
+            "type": "string",
+            "description": "The name of the hybrid machine."
+          },
+          {
+            "name": "extensionUpgradeParameters",
+            "in": "body",
+            "required": true,
+            "schema": {
+              "$ref": "#/definitions/MachineExtensionUpgrade"
+            },
+            "description": "Parameters supplied to the Upgrade Extensions operation."
+          }
+        ],
+        "responses": {
+          "200": {
+            "description": "OK"
+          },
+          "202": {
+            "description": "HTTP 202 (Accepted) if the operation was successfully started and will complete asynchronously."
+          },
+          "default": {
+            "description": "Error response describing why the operation failed.",
+            "schema": {
+              "$ref": "../../../../../common-types/resource-management/v2/types.json#/definitions/ErrorResponse"
+            }
+          }
+        },
+        "x-ms-examples": {
+          "Upgrade Machine Extensions": {
+            "$ref": "./examples/Extensions_Upgrade.json"
+          }
+        },
+        "x-ms-long-running-operation": true
+      }
+    },
+    "/providers/Microsoft.HybridCompute/location/{location}/publisher/{publisher}/extensionType/{extensionType}/version/{version}": {
```

Comment:
>version [](http://example.com/codeflow?start=121&length=7)

Has this api been reviewed internally ? 

This API has a number of issues 

- You seem to be retuning a list of resources for a point get 
- The resources are not described as ARM resources.
- The parent resources (publisher/{publisher}) for the child resource has not been defined. 
- You seem to be using the URI definition for filtering on properties rather than defining specific resources and using ODATA filtering mechanisms 

Can you please restructure this api according to ARM guidelines and resubmit the PR ?

Reply:
Hi, this api has been reviewed within the team. To address your concerns :

- We are modeling after an existing API- (https://docs.microsoft.com/en-us/powershell/module/az.compute/get-azvmextensionimage?view=azps-8.2.0). Within this, please see the operatiosnstatus api which returns a list of operations off a point get.
- Please clarify what you mean by ARM Resources. We currently have a manifest PR out right now : https://armmanifest.visualstudio.com/One/_git/Manifest/pullrequest/5501 
- Publisher and Type go as a single entity to identify the Extension. So cannot define just the publisher as a parent resource
- The objective of this API would be to filter based on the properties provided above. For example GET publisher+Type will provide a response of all the potential versions.

Reply:
+1 to @raosuhas comments above. The ARM resource model requires both collection and singleton GET endpoints. Collection GET is on the resource type, singleton GET is on an individual resource. Ref: https://armwiki.azurewebsites.net/api_contracts/guidelines/rpc.html#rpc006-tracked-resource-type-supports-get-put-patch-delete--list-at-all-scopes-as-applicable.

ARM resources must have top-level name, id, and type properties at a minimum (proxy resources). Tracked resources also have location and tags. Ref: https://armwiki.azurewebsites.net/api_contracts/guidelines/rpc.html#rpc016-incorrect-resource-requests-and-responses. RP-specific properties are contained with in a property bag. Ref: https://armwiki.azurewebsites.net/api_contracts/guidelines/rpc.html#rpc013-resource-definition-must-define-a-property-bag

Reply:
@mentat9 
What is the definition of a proxy resource? Does `/subscriptions/{subscriptionId}/providers/Microsoft.Compute/locations/{location}/publishers/{publisherName}/artifacttypes/vmextension/types/{type}/versions` count as a proxy resource?
We modeled our API based on the Compute API. I looked at their most recent stable API swagger, it seemed like they didn't meet the guideline either https://github.com/Azure/azure-rest-api-specs/blob/208c91dc71eec634400fb0c0a7e073e35afd2978/specification/compute/resource-manager/Microsoft.Compute/ComputeRP/stable/2022-03-01/virtualMachineExtensionImage.json#L156

Reply:
No, that URI specifies the collection of instances of the "versions" resource type. More info on proxy resources here: https://armwiki.azurewebsites.net/rp_onboarding/tracked_vs_proxy_resources.html.

That ComputeRP spec returns a collection of instances of the "versions" type, which is the correct behavior (and members are spec'd as proxy resources). Unfortunately, they messed up the response payload layout, which never should be just an array. The correct response is an object with a "value" property that is an array of resources and [optionally] a nextLink string property. Ref: https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/resource-api-reference.md#pagination and https://armwiki.azurewebsites.net/api_contracts/guidelines/openapi.html#oapi009-always-add-x-ms-pageable-to-list-calls.

A difference in your spec from the ComputeRP link you provided is that the collection response in your spec was on the individual GET Uri (.../version/{version}), not the resource type (.../versions) as spec'd by ComputeRP. Note also the type name "version" should always instead be a plural ("versions") as ComputeRP did. Your resource type needs both an individual GET and a collection GET (as ComputeRP has). You also need to provide APIs for the parent resource types.

Note: to respond to ARM reviewers, remove the ARMChangesRequested label which will automatically notify the current oncall that your PR needs another look. **This is important: otherwise ARM reviewers will not know your PR is ready.** It's Your PR will be

Reply:
@mentat9 in the CRP sample above, "locations/publishers" is a proxy resource type defined in [the CRP manifest] (https://armmanifest.visualstudio.com/One/_git/Manifest?path=/MICROSOFT.COMPUTE/PROD/MANIFEST.JSON&version=GBmaster&line=33754&lineEnd=33754&lineStartColumn=15&lineEndColumn=37&lineStyle=plain&_a=contents).  How do I understand the types and versions? Are they resource type too? If so, why don't they have to define them in manifest?

Reply:
Yes, that is a resource type. The naming of resource types follows a shorthand, so locations/publishers type would have a collection Uri of ```.../Microsoft.Compute/locations/{locationName}/publishers``` and a singleton Uri of ```.../Microsoft.Compute/locations/{locationName}/publishers/{publisherName}```

That link **is** the definition of a resource type in the manifest.

Reply:
Thanks @mentat9.  The  `types` and `versions` in the CRP sample below are also resource type? `/subscriptions/{subscriptionId}/providers/Microsoft.Compute/locations/{location}/publishers/{publisherName}/artifacttypes/vmextension/types/{type}/versions`

Reply:
> Thanks @mentat9. The `types` and `versions` in the CRP sample below are also resource type? `/subscriptions/{subscriptionId}/providers/Microsoft.Compute/locations/{location}/publishers/{publisherName}/artifacttypes/vmextension/types/{type}/versions`

@jianyunt - That path specification identifies 5 resource types and 4 resource instances. Resource instances are identified by their resource type and a parameter or hard-coded segment.

**Fully qualified resource types referred to in that path:**
```
Microsoft.Compute/locations
Microsoft.Compute/locations/publishers
Microsoft.Compute/locations/publishers/artifacttypes
Microsoft.Compute/locations/publishers/artifacttypes/types
Microsoft.Compute/locations/publishers/artifacttypes/types/versions
```

**Resource instances referred to in that path:**
```
Microsoft.Compute/locations/{location}
Microsoft.Compute/locations/{location}/publishers/{publisherName}
Microsoft.Compute/locations/{location}/publishers/{publisherName}/artifacttypes/vmextension
Microsoft.Compute/locations/{location}/publishers/{publisherName}/artifacttypes/vmextension/types/{type}
```

Since it ends with a resource type segment, the overall path spec identifies the collection of all instances of the ```Microsoft.Compute/locations/publishers/artifacttypes/types/versions``` resource type.